doc/proofgeneral.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.29.
.\" Well, actually, it was edited after that by me, so the dire
.\" warning should not be greatly feared.
.\" This is based on the output of: help2man ../bin/proofgeneral
.\"
.\" $Id$
.\"
.TH PROOFGENERAL "1" "March 2004" "proofgeneral" "User Commands"
.SH NAME
proofgeneral \- launch Emacs Proof General
.SH SYNOPSIS
.B proofgeneral
[\fIOPTION\fR] [\fIFILE\fR]...
.SH DESCRIPTION
Launches Emacs Proof General, editing the proof script FILE.
.SH OPTIONS
.TP
\fB\-\-emacs\fR
startup Proof General with emacs (GNU Emacs)
.TP
\fB\-\-xemacs\fR
startup Proof General with xemacs (XEmacs)
.TP
\fB\-\-emacsbin\fR <EMACS>
startup Proof General with emacs binary <EMACS>
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help and exit
.TP
\fB\-v\fR, \fB\-\-version\fR
output version information and exit
.TP
Unrecognized options are passed to Emacs, along with file names.
.SH EXAMPLES
.TP
proofgeneral Example.thy
Load Proof General editing Isar file Example.thy
.TP
proofgeneral example.v
Load Proof General editing Coq file Example.v
.PP
For documentation and latest versions, visit http://proofgeneral.inf.ed.ac.uk.
.SH AUTHOR
David Aspinall.
.SH "REPORTING BUGS"
Report bugs to <da+pg-bugs@inf.ed.ac.uk>.
.SH COPYRIGHT
Copyright \(co 1998-2004 LFCS, University of Edinburgh, UK.
.br
This is free software; see the source for copying conditions.
.SH "SEE ALSO"
The full documentation for Proof General is available within
the program, and also as the Texinfo manual.  This should be
available by the command
.IP
.B info ProofGeneral
.PP
if the info file has been installed.  There is a second
manual containing instructions on adapting Proof General
to new proof assistants:
.IP
.B info PG-adapting
.PP
Ordinary users do not need to consult this manual.


NEWS
v1.15 (PRERELEASE)

Extension

* Now builds for Firefox using WebExtension hooks
* Use chrome.notifications instead of webkitNotifications
* Use <div> with style instead of inline <img> for edit button

edit-server.el

* avoid calling make-frame-on-display for TTY setups
* restore edit-server-default-major-mode if auto-mode lookup fails

v1.14

Extension

* Fix a bug with the context menu (#85)
* Make text buttons slightly opaque

edit-server.el

* add edit-server-edit-mode-hook (#91, #94)

v1.13

Extension

* Change the handling of hidden elements (fix bug #78)
* Add debugging for erroneous hidden text areas (#93)
* Add keyboard shortcut to bring Emacs to foreground
* Pass clipboard contents to foreground request

edit-server.el

* add advice to save-buffers-kill-emacs to avoid prompting on shutdown
* add autoload cookies
* fix bug with format chars in url (#80)
* don't call kill buffer hooks twice (#92)
* don't set-buffer-multibyte on process buffer
* support the "foreground" request with optional clipboard contents

v1.12

Extension

* Upgraded manifest to version 2 for future releases of Chrome
* Added an "Enable Debug" flag to control logging to console
* Use Mutation Observers to better track changes in dynamic DOMs
* Fake a keypress when updating the textarea
* Bug fix for handling editable DIV elements

edit-server.el

* Now available from the MELPA package archive!
* Beta support for iterative C-x C-s saving (it works but is ugly, see kill-ring)
* Default to using UTF-8 coding for process communication
* Raise existing frame consistently on Mac
* Tweak the portion of the URL used to name buffers

v1.11

Extension

* Fixed some breakage to the "edit focused area" feature
* Fixed broken link to embedded edit-server

edit-server.el

* A couple of fixes for Emacs 24 compatibility
* Globalize the minor mode so it persists through major-mode changes

v1.10

Extension

* Ignore textareas marked as read only
* Don't tag areas that are not visible
* General clean-up to use jQuery to find elements
* Explicit CSS for edit button to override page settings
* Handle editable <div> blocks (e.g. Google+)
* Optimise the finding of text areas for highly dynamic pages
* Revamp the settings page with "Fancy Settings"

edit-server.el

* Allow customisation of edit-server-default-major-mode
* Allow edit mode to be set by matched URL
* Tweak detection of MacOS X Emacsen
* Change behaviour of C-x C-s to save to kill-ring
* Persist the buffer-local variables beyond mode changes
* Setup keymap within defvar
* Clean-ups to code to be more idiomatic.

v1.9

Extension

* Support for iterative editing (python server only currently)
* Add context menu while in text area box
* Fix reference to Alt-E, it's actually Alt-Enter
* Trigger a DOM change() event when we update the text area

edit-server.el

* Actively encode responses as UTF-8
* Make edit-server-edit-mode a minor mode instead of a derived mode

v1.8

Extension

* Added option to enable/disable visual edit boxes
* Improved feedback as editable elements come in and out of focus
* Updated text box will now fade from yellow after an update

edit-server.el

* Added edit-server-start-hook for additional customisation when edit starts

v1.7

Extension

* New icon state. Blue=Waiting, Green=In Progress, Red=Error
* Improved mouse-over text for icon to give more useful feedback

edit-server.el

* Move all frame customisation into edit-server-new-frame-alist
* Don't ask user before closing emacs and network process
* Just skip creating new network process if it's already running
* Make sure edit buffer is selected on new frames
* Prompt window manager to bring new frames to the top of the stack (X windows only)
* Enable multi-byte mode on edit buffers for better unicode handling
* Explicitly fail on XEmacs if make-network-process isn't found (XEmacs patches welcome)

v1.6

Extension

* Send the page URL to the edit server via x-url header
* Options page: direct people to Emacs Wiki page for gory extras
* Make Double Click to Edit a selectable option
* Make Keyboard Shortcut to Edit a selectable option
* Don't pass edit ID in URL, but in header as x-id (like x-url)

edit-server.el

* Fix edit-server-kill-client (passed process not buffer)
* Fix unkillable zombie buffers created by a closed browser
* Name buffer based on x-url if it exists
* Workaround make-frame-on-display bug on Aquamacs

v1.5

Extension

* Clean-up and streamline the text tagging code
* Process sub FRAMEs and IFRAMEs looking for text boxes
* Handle textareas that have been "cloned" by Javascript - fixes Google Groups
* Generate user feedback if "edit focused area" failed
* Skip textareas with spell-checking disabled (usually email address boxes)
* Make Alt-Enter in focused area generate an edit request (patches to make it configurable welcome!)
* Updates to help text in Options

edit-server.el

* Allow more customisation of the pop-up frame
* Allow optional binding to a non-local IP address
* Updates to help text in server code.

v1.4

Extension

* Added proper Emacs icon for extension
* Added options page with instructions and option to change port
* Allow a simple status test to be run in the options page
* Change mouse cursor when hovering over edit button
* Double click on a text area invokes the editor

edit-server.el

* edit-server.el now distributed in-situ with the extension
* Renamed edit_server to edit-server to follow conventions
* Stopped C-x C-c from shutting down session (leaving stale TEXTAREA)
* Added default size and name to new frames
* Clear modification indication on new edit buffers


images/gimp/scripts/proofgeneral.scm
;;
;; Gimp script fu to make buttons from a source .xcf file.
;;
;; David Aspinall <da@dcs.ed.ac.uk>
;;
;; $Id$
;;


;; TODO: make greyed out, pressed, unpressed versions.
;; e.g. : Add bevel for "up" position:
;; (script-fu-add-bevel 0 image
;;     (car (gimp-image-active-drawable image)) "10" 0 0)

(define (script-fu-proofgeneral-make-button buttonname)
  (let* ((filename (string-append buttonname ".xcf"))
	 (image    (car (gimp-file-load 1 filename filename)))
	 (xpmname  (string-append buttonname ".xpm"))
	 (poor-xpm (string-append buttonname ".8bit.xpm"))
	 (xbmname  (string-append buttonname ".xbm")))
    (gimp-image-flatten image)
    ;; Full xpm
    (gimp-file-save 1 image (car (gimp-image-active-drawable image))
		    xpmname xpmname)
    ;; Impoverised xpm
    (gimp-convert-indexed image 1 0 8 1 1 "blah")
    (gimp-file-save 1 image (car (gimp-image-active-drawable image))
		    poor-xpm poor-xpm)
    ;; Mono image
    (gimp-convert-rgb image)
    (gimp-image-flatten image)
    (gimp-convert-indexed image 1 1 3 2 1 "blah")
    (file-xbm-save 1 image (car (gimp-image-active-drawable image))
		    xbmname xbmname
		    "Proof General button"
		    FALSE 0 0 "")
    ;; Finish
    (gimp-image-delete image)
    ))

(script-fu-register "script-fu-proofgeneral-make-button" 
		    "<Toolbox>/Xtns/Script-Fu/Proof General/Make Button"
		    "Save buttons in various formats"
		    "da@dcs.ed.ac.uk" "da@dcs.ed.ac.uk"
		    "1998/10/04"
		    ""
		    SF-VALUE "Button/file name" "\"goal\"")

(define (script-fu-proofgeneral-make-all-buttons)
  (mapcar script-fu-proofgeneral-make-button
	  '("goal" "next" "qed" "restart" "retract" "undo" "use" "state" "context" "info" "command" "find" "help" "interrupt" "goto")))

(script-fu-register "script-fu-proofgeneral-make-all-buttons" 
		    "<Toolbox>/Xtns/Script-Fu/Proof General/Make All Buttons"
		    "Save Proof General buttons in the various formats"
		    "da@dcs.ed.ac.uk" "da@dcs.ed.ac.uk"
		    "1998/10/04"
		    "")

(define (script-fu-proofgeneral-save-pic imgname)
  (let* ((filename (string-append imgname ".xcf"))
	 (image    (car (gimp-file-load 1 filename filename)))
	 (jpgname  (string-append imgname ".jpg"))
	 (gifname  (string-append imgname ".gif"))
	 (poorgifname  (string-append imgname ".8bit.gif")))
    ;; Flatten and save as jpg
    ;;(gimp-image-flatten image)
    ;; Flattening forces a white background.  Let's use merge.
    (if (> (car (gimp-image-get-layers image)) 1)
	(gimp-image-merge-visible-layers image 0))
    (file-jpeg-save 1 image (car (gimp-image-active-drawable image))
		    jpgname jpgname
		    0.75 0 1)
    ;; gif with full palette
    (gimp-convert-indexed image TRUE 255)
    (file-gif-save 1 image (car (gimp-image-active-drawable image))
		    gifname gifname
		    FALSE FALSE 0 0)
    ;; gif with impoverished palette for display in XEmacs
    (gimp-convert-rgb image)
    (gimp-convert-indexed image 1 15)
    (file-gif-save 1 image (car (gimp-image-active-drawable image))
		    poorgifname poorgifname
		    FALSE FALSE 0 0)
    ;; Finish
    (gimp-image-delete image)
    ))

(script-fu-register "script-fu-proofgeneral-save-jpg" 
		    "<Toolbox>/Xtns/Script-Fu/Proof General/Save Jpeg"
		    "Save image as jpeg"
		    "da@dcs.ed.ac.uk" "da@dcs.ed.ac.uk"
		    "1998/10/04"
		    ""
		    SF-VALUE "Basename" "\"test\"")


(define (script-fu-proofgeneral-save-all-pix)
  (mapcar script-fu-proofgeneral-save-pic
	  '("ProofGeneral" "pg-text")))

(script-fu-register "script-fu-proofgeneral-save-all-jpegs" 
		    "<Toolbox>/Xtns/Script-Fu/Proof General/Save all Jpegs"
		    "Save Proof General images as jpegs"
		    "da@dcs.ed.ac.uk" "da@dcs.ed.ac.uk"
		    "1998/10/04"
		    "")